### PR TITLE
Random Number: Fixed #3168 and #3169, limit triggers to only when you have two different numbers

### LINKED ISSUE #3168: Random Number: Prevent triggering on just 'random'
See ['random'](https://duckduckgo.com/?q=random&ia=answer) - we shouldn't trigger on such a generic query.

---

IA Page:  http://duck.co/ia/view/random_number 
[Maintainer](http://docs.duckduckhack.com/maintaining/guidelines.html): @jonk1993

### PR DESCRIPTION
**Pull request by jmwoliver**
###### What does your Pull Request do (check all that apply)?

Choose the most relevant items and use the following title template to name
your Pull Request.
- [ ] New Instant Answer
  - [ ] Cheat Sheets: **`New {Cheat Sheet Name} Cheat Sheet`**
  - [ ] Other: **`New {IA Name} Instant Answer`**
- [X] Improvement
  - [X] Bug fix: **`Random Number: Fix #3168 and #3169`**
  - [ ] Enhancement: **`{IA Name}: {Description of Improvements}`**
- [ ] Non–Instant Answer
  - [ ] Other (Role, Template, Test, Documentation, etc.): **`{GoodieRole/Templates/Tests/Docs}: {Short Description}`**
###### Description of changes

Limit triggers to only when you have two different numbers. Don't trigger if lower and upper bounds are the same, there is only one number, or no numbers present.
###### Which issues (if any) does this fix?

Fixes #3168 and #3169 - Limit triggers to only when you have two different numbers
###### People to notify (@mention interested parties)

@GuiltyDolphin 

---

Instant Answer Page: https://duck.co/ia/view/random_number

[Maintainer](http://docs.duckduckhack.com/maintaining/guidelines.html): @jonk1993

### CODE REVIEW

**Comment by daxtheduck** [comment]
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

**Comment by jmwoliver** [comment]
@GuiltyDolphin the IA page may need to be changed because it says 'Random Number' is a viable trigger.

File: lib/DDG/Goodie/RandomNumber.pm [hunk 1]
```diff
@@ -17,6 +17,7 @@ handle query_lc => sub {
 
     my $start = $1 || 0;
     my $end   = $2 || 0;
+    return unless ($start != $end) && ($end != undef);
```
Comment (jonk1993) [inline lib/DDG/Goodie/RandomNumber.pm]: So I think this may have to be handled a bit differently. The lines 

```
 my $start = $1 || 0;
 my $end   = $2 || 0;
```

Will guarantee that `$start` and `$end` will be set to `0` if their regex groups are not matched. So `($end != undef)` will never be true, as `$end` will default to `0`.

Since we aren't including negatives as valid input, one possibility is to change `$start` and `$end` initialization to:

```
 my $start = $1 || -1;
 my $end   = $2 || -1;
```

and change your return statement to return unless `$end` does not equal `-1`.

**Comment by GuiltyDolphin** [comment]
@jmwoliver Yes - `random number` *is* a valid trigger; it is `random` on its own that I'm concerned about.

File: lib/DDG/Goodie/RandomNumber.pm (hunk 1 above)
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/RandomNumber.pm]: I think just doing `return if ($start == $end)` should be fine - we already ensure they aren't undefined (as @jonk1993).

Btw, @jonk1993 why are we using `query_lc` rather than `remainder_lc`? Assuming the triggerword is `random` we can just check the remainder for `number`.

**Comment by jmwoliver** [comment]
@jonk1993 Okay, that makes sense, I'll try it that way. @GuiltyDolphin Thank you for clarifying. I was wondering, would "random 1 10" still be a valid input or should it only trigger on "random number 1 10"?

**Comment by GuiltyDolphin** [comment]
@jmwoliver I think `random 1 10` should still trigger - the cases that should not trigger are:

1. `random` (on its own)
2. Anything that causes the lower and upper bounds to be the same.

**Comment by jmwoliver** [comment]
@GuiltyDolphin It should be good now. It fails on just 'rand'/'random' being passed in, and it fails when only one number is given and when both numbers given are the same. And it still works if you just put 'random number'. It also still works with 'random 1 10'. It looks like some test cases failed, I'll give it a look.

File: lib/DDG/Goodie/RandomNumber.pm [hunk 2]
```diff
@@ -30,14 +37,17 @@ handle query_lc => sub {
     ($end, $start) = ($start, $end) if ($start > $end);
 
     my $valDiff = $end - $start;
+    
+    #returns if both end and start are equal, and end has been changed by the user
+    return if ($valDiff == 0) && ($end != 0);
```
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/RandomNumber.pm]: Why do we have so many checks? One check of `return if $start == $end` should suffice.

File: lib/DDG/Goodie/RandomNumber.pm [hunk 3]
```diff
@@ -30,14 +37,17 @@ handle query_lc => sub {
...
 
     my $rand = rand;
 
     if ($start && $end || $valDiff > 1) {
         $rand *= ($valDiff + 1);
         $rand = int($rand) + $start;
     }
-
+
```
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/RandomNumber.pm]: Please ensure you don't add any unnecessary whitespace.

File: lib/DDG/Goodie/RandomNumber.pm [hunk 4]
```diff
@@ -13,10 +13,17 @@ handle query_lc => sub {
...
+    # checks if only 'rand' or 'random' is entered
+    return if ($_ =~ /^\!?(?:rand(?:om|))$/i);
+    
     return unless ($_ =~ /^\!?(?:rand(?:om|)(?: num(?:ber|)|)(?: between|))( [\d\.]+|)(?: and|)( [\d\.]+|)$/i);
 
     my $start = $1 || 0;
     my $end   = $2 || 0;
+    
+    #makes sure user inputs two numbers
+    return if ($start != $end) && ($end == 0);
```
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/RandomNumber.pm]: This will be caught by the regex.

File: lib/DDG/Goodie/RandomNumber.pm [hunk 5]
```diff
@@ -13,10 +13,17 @@ handle query_lc => sub {
     srand();
     # Random number.
     # q_check (as opposed to q_internal) Allows for decimals.
+    
+    # checks if only 'rand' or 'random' is entered
+    return if ($_ =~ /^\!?(?:rand(?:om|))$/i);
+    
     return unless ($_ =~ /^\!?(?:rand(?:om|)(?: num(?:ber|)|)(?: between|))( [\d\.]+|)(?: and|)( [\d\.]+|)$/i);
```
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/RandomNumber.pm]: I think this regex is causing some issues, if we clean it up we shouldn't need the checks for additional numbers - having `blah|` in each case it allowing too many invalid forms (for example, if `and` is present, we would *expect* there to be a second number).

**Comment by jmwoliver** [comment]
@GuiltyDolphin I took out the spaces and recommitted and Travis likes it now for some reason. The reason I can't just do 'return if ($start == $end)' is because both are initialized to 0 if not given a value. So if you only search "random number" which is a valid search, then $start == $end and that return statement would not trigger this IA. You are probably right that there is a better way to do it, so I can change it if you'd like. But the way I have it right now fulfills all the parameters for this IA.

File: t/RandomNumber.t [hunk 6]
```diff
@@ -36,6 +36,13 @@ ddg_goodie_test(
     'random number between 0 and 1'   => build_test(0, 1, qr/$RE{num}{real}/),
     'random number between 0 and 10'  => build_test(0, 10, qr/\d{1,2}/),
     'random number between 0 and 100' => build_test(0, 100, qr/\d{1,3}/),
+
+    #cases that don't have any numbers, only one number, or both numbers the same should fail
+    'random number between 1 and '    => undef,
+    'random 10'                       => undef,
+    'random 7 7'                      => undef,
+    'random'                          => undef,
```
Comment (GuiltyDolphin) [inline t/RandomNumber.t]: Lets add a test for `random 0 0`.

**Comment by jmwoliver** [comment]
@GuiltyDolphin What is the desired behavior for that? Right now it's finding a random between 0 and 1 but outputting 0 - 0. Obviously that is very wonky and not what we want.
aoeu

**Comment by GuiltyDolphin** [comment]
@jmwoliver It falls into the case of upper and lower bounds being the same (though it seems like another bug has unearthed itself); so it should not trigger.

**Comment by jmwoliver** [comment]
@GuiltyDolphin  I just added `return if ($start == $end) && ($end == 0);` and it fixed the issue. I will commit it with the new test case.

**Comment by jmwoliver** [comment]
@GuiltyDolphin Okay I think it is finally good to go. My solution was really make-shift though by adding `return if ($start == $end) && ($end == 0) && !(($_ =~ /^\!?(?:rand(?:om|)(?: num(?:ber|)|))$/i));` so it would fail if both numbers entered are 0 but not if you just type 'random number'.

File: lib/DDG/Goodie/RandomNumber.pm [hunk 7]
```diff
@@ -13,10 +13,18 @@ handle query_lc => sub {
...
 
     my $start = $1 || 0;
     my $end   = $2 || 0;
+    
+    #makes sure user inputs two numbers, and both numbers are not zero
+    return if ($start != $end) && ($end == 0);
+    return if ($start == $end) && ($end == 0) && !(($_ =~ /^\!?(?:rand(?:om|)(?: num(?:ber|)|))$/i));
```
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/RandomNumber.pm]: I think it might be best just to use the `remainder` handler - then we can just check `matched_trigger` and `remainder` to see what sort of query should be handled.

File: lib/DDG/Goodie/RandomNumber.pm (hunk 1 above)
Comment (jonk1993) [inline lib/DDG/Goodie/RandomNumber.pm]: Ah, I'm not the creator just became the maintainer a week or so ago. Changing to `remainder_lc` seems to makes sense.

**Comment by jmwoliver** [comment]
@GuiltyDolphin Could you explain further on how to use `remainder_lc` or point me to an IA that uses it how you are suggesting?

**Comment by GuiltyDolphin** [comment]
@jmwoliver I'd reccommend the [docs](http://docs.duckduckhack.com/backend-reference/handle-functions.html).

**Comment by jmwoliver** [comment]
@GuiltyDolphin Sorry, I still haven't figured out how to implement `remainder_lc`. I'll keep looking at it but I couldn't figure it out from the docs, is there anything else you could provide?

**Comment by GuiltyDolphin** [comment]
@jmwoliver There's not much to it - just have a look at some existing Goodies and how they use the handlers in different ways; the docs do a fairly thorough job of explaining though.

**Comment by zekiel** [comment]
@jmwoliver  still need some help?

**Comment by jmwoliver** [comment]
@zekiel Yes please. I still couldn't figure it out by looking at the documentation and examples.

**Comment by gaulrobe** [comment]
@jmwoliver 

Right now, the handler being used is `query_lc` which will give you the complete query that was used to trigger the IA. So if the query is `number between 0 and 10` the incoming value you have to work with (`$_`) will be `number between 0 and 10`.

If you switch the handler to `remainder_lc` you will get the query minus the trigger. So in the above example, `$_` would be `between 0 and 10`. What this means is that you don't have to do additional trigger validation. Just validation on the rest of the query. This _should_ simplify the regex being used.

Hope this helps!

**Comment by mintsoft** [comment]
@jmwoliver Hiya, just checking in. Are you OK continuing with this? If not we can open it up to the community :+1:

**Comment by moollaza** [comment]
This PR is being closed due to inactivity. If you're still interested in finishing, please let me know and I'll gladly re-open it for you!
